### PR TITLE
isc: fix foms hashing by localities

### PR DESCRIPTION
**Pull request by andriytk**
Currently, all computation requests foms for an object
are assigned to the same locality, which kills parallelism
of the computation. Even when we change hashing by object
id to the cob fid at fom_home_locality(), it does not help.
Because the hashing function works very badly.

Solution: 1) use cob fid for hashing instead of object id
and 2) change hashing function to the improved version of
well-known djb2 hash-function.

### CODE REVIEW

File: lib/hash.c [hunk 1]
```diff
@@ -276,24 +276,34 @@ M0_INTERNAL uint64_t m0_htable_size(const struct m0_htable *htable)
...
+
+	for (p = buf; n-- > 0; p++) {
+		hash += ((unsigned long long)*p << 32) | *p;
+		hash *= 2971215073;
+		hash >>= 16;
+		hash *= 2971215073;
+	}
+
+	return hash;
+}
```
Comment (nikitadanilov) [inline lib/hash.c:302]: Ah, polynomial hashes. It's easy to check that the value of hash(p, n) is (modulo 2^64) `p[0]*a*b^n + p[1]*a*b^{n-1} + ... p[n-1]*a*b` for certain a and b. Therefore, if you have some string R, such that `hash(R) == 0`, then for any string X, `hash(R + X) == hash(R + R + X) == ... == 0`, where + is string catenation. That's not very secure.

**Review by nikitadanilov — CHANGES_REQUESTED** [review]

File: lib/hash.c (hunk 1 above)
Comment (andriytk) [reply lib/hash.c:302]: If we need security here, we should probably use SHA-256 or something.

Comment (huanghua78) [reply lib/hash.c:302]: Please make sure this piece of code is not just copied from somewhere.
This might violate the copyright.

Comment (andriytk) [reply lib/hash.c:302]: @huanghua78 It is not. The only line which is copied (from http://www.cse.yorku.ca/~oz/hash.html) is `uint64_t hash = 5381;` and even it has changed type. Anyway, it seems the original algorithm is in public domain, so I think we are safe.

Comment (huanghua78) [reply lib/hash.c:302]: OK. Thank you.

**Review by nikitadanilov — APPROVED** [review]

**Comment by andriytk** [comment]
@madhavemuri could you please review it also?

File: iscservice/isc.c [hunk 2]
```diff
@@ -94,9 +94,16 @@ struct m0_sm_state_descr isc_fom_phases[] = {
...
+	size_t             hash;
+
+	M0_ENTRY("fom=%p cob="FID_F, fom, FID_P(&fi->fi_cob));
+
+	hash = m0_fid_hash(&fi->fi_cob);
+
+	M0_LEAVE("hash=%zx", hash);
 
-	return m0_fid_hash(&comp_req->icr_comp_fid);
+	return m0_fid_hash(&fi->fi_cob);
```
Comment (madhavemuri) [inline iscservice/isc.c:106]: "return hash;" from here instead of additional m0_fid_hash().

Comment (andriytk) [reply iscservice/isc.c:106]: @madhavemuri, sorry - I'm not sure I understand that. Which hash?

Comment (huanghua78) [reply iscservice/isc.c:106]: The 'hash‘ is already computed and stored in this variable.
So, as @madhavemuri said, we can have:
```
        return hash;
```

Comment (andriytk) [reply iscservice/isc.c:106]: Ah, sorry - I just missed the context. Yes, you are right. Will fix it now.

**Review by huanghua78 — APPROVED** [review]
Looks fine.

**Comment by shailesh-vaidya** [comment]
<h1>Jenkins CI Result : <a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615/>Motr#615</a></h1><h2>Motr Test Summary</h2><table><tr><th>Test Result</th><th>Count</th><th>Info</tr><tr><td>:x:Failed</td><td>7</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/00userspace-tests target=_blank>01motr-single-node/00userspace-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/01kernel-tests target=_blank>01motr-single-node/01kernel-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/28motr-sys-kvs-test target=_blank>01motr-single-node/28motr-sys-kvs-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/57sns-repair-motr-1k-1f target=_blank>03motr-single-node/57sns-repair-motr-1k-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/34sns-repair-1n-1f target=_blank>04motr-single-node/34sns-repair-1n-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/52singlenode-sanity target=_blank>01motr-single-node/52singlenode-sanity</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/51kem target=_blank>02motr-single-node/51kem</a></p></details></td></tr><tr><td>:checkered_flag:Skipped</td><td>5</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/37protocol target=_blank>01motr-single-node/37protocol</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/31motr-sys-kvs-kernel-test target=_blank>04motr-single-node/31motr-sys-kvs-kernel-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/07multi-client-tests target=_blank>05motr-single-node/07multi-client-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/12sns-cc-io-tests target=_blank>05motr-single-node/12sns-cc-io-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/39mount-fail target=_blank>05motr-single-node/39mount-fail</a></p></details></td></tr><tr><td>:heavy_check_mark:Passed</td><td>52</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/03initscript-tests target=_blank>01motr-single-node/03initscript-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/04net-tests target=_blank>01motr-single-node/04net-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/05rpcping-tests target=_blank>01motr-single-node/05rpcping-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/06console-tests target=_blank>01motr-single-node/06console-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/29confgen-test target=_blank>01motr-single-node/29confgen-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/30hagen-test target=_blank>01motr-single-node/30hagen-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/35m0singlenode target=_blank>01motr-single-node/35m0singlenode</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/01motr-single-node/46m0crate target=_blank>01motr-single-node/46m0crate</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/02system-tests target=_blank>02motr-single-node/02system-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/08degraded-mode-tests target=_blank>02motr-single-node/08degraded-mode-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/09poolmach-tests target=_blank>02motr-single-node/09poolmach-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/20pool-version-test target=_blank>02motr-single-node/20pool-version-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/21spiel-test target=_blank>02motr-single-node/21spiel-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/22rpc-cancel-test target=_blank>02motr-single-node/22rpc-cancel-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/23m0d-signal-test target=_blank>02motr-single-node/23m0d-signal-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/24m0d-fsync-test target=_blank>02motr-single-node/24m0d-fsync-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/41motr-conf-update target=_blank>02motr-single-node/41motr-conf-update</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/47motr-tests-user-kernel target=_blank>02motr-single-node/47motr-tests-user-kernel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/02motr-single-node/53clusterusage-alert target=_blank>02motr-single-node/53clusterusage-alert</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/16conf-st-test target=_blank>03motr-single-node/16conf-st-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/17sss-st-test target=_blank>03motr-single-node/17sss-st-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/36spare-reservation target=_blank>03motr-single-node/36spare-reservation</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/54sns-repair-motr-1f target=_blank>03motr-single-node/54sns-repair-motr-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/55sns-repair-motr-1n-1f target=_blank>03motr-single-node/55sns-repair-motr-1n-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/56sns-repair-motr-mf target=_blank>03motr-single-node/56sns-repair-motr-mf</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/58sns-repair-motr-ios-fail target=_blank>03motr-single-node/58sns-repair-motr-ios-fail</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/59sns-repair-motr-abort target=_blank>03motr-single-node/59sns-repair-motr-abort</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/60sns-repair-motr-abort-quiesce target=_blank>03motr-single-node/60sns-repair-motr-abort-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/03motr-single-node/61spiel-multi-confd target=_blank>03motr-single-node/61spiel-multi-confd</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/19sns-abort-test target=_blank>04motr-single-node/19sns-abort-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/25spiel-sns-repair-test target=_blank>04motr-single-node/25spiel-sns-repair-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/26spiel-sns-repair-quiesce-test target=_blank>04motr-single-node/26spiel-sns-repair-quiesce-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/27sns-repair-io-fail-test target=_blank>04motr-single-node/27sns-repair-io-fail-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/32m0t1fs-rconfc-fail-test target=_blank>04motr-single-node/32m0t1fs-rconfc-fail-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/33motr-st target=_blank>04motr-single-node/33motr-st</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/48motr-raid0-io target=_blank>04motr-single-node/48motr-raid0-io</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/49motr-rpc-cancel target=_blank>04motr-single-node/49motr-rpc-cancel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/50motr-rm-lock-cc-io target=_blank>04motr-single-node/50motr-rm-lock-cc-io</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/04motr-single-node/51motr-rmw target=_blank>04motr-single-node/51motr-rmw</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/10sns-single-tests target=_blank>05motr-single-node/10sns-single-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/11sns-multi-tests target=_blank>05motr-single-node/11sns-multi-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/13sns-repair-quiesce target=_blank>05motr-single-node/13sns-repair-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/14m0t1fs-fsync-test target=_blank>05motr-single-node/14m0t1fs-fsync-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/15m0t1fs-fwait-test target=_blank>05motr-single-node/15m0t1fs-fwait-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/18m0mount-test target=_blank>05motr-single-node/18m0mount-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/38sns-abort-quiesce-test target=_blank>05motr-single-node/38sns-abort-quiesce-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/40motr-dgmode target=_blank>05motr-single-node/40motr-dgmode</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/42motr-utils target=_blank>05motr-single-node/42motr-utils</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/43motr-sync-replication target=_blank>05motr-single-node/43motr-sync-replication</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/44motr-sns-repair target=_blank>05motr-single-node/44motr-sns-repair</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/05motr-single-node/45motr-sns-repair-N-1 target=_blank>05motr-single-node/45motr-sns-repair-N-1</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit/motr-single-node/62dix-repair-st-lookup-insert target=_blank>motr-single-node/62dix-repair-st-lookup-insert</a></p></details></td></tr><tr><td>Total</td><td>64</td><td><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/615//testReport/junit>:link:</a></td></tr></table><h2> CppCheck Summary</h2><p> &emsp;&emsp;&emsp;Cppcheck: No new warnings found :thumbsup:</p>

**Comment by andriytk** [comment]
Oh, this should have not been merged before the test results... Let me check if it can be fixed quickly. If not, I will create a revert PR.

**Pull request merged**